Add feature of refreshing converted files if original files are modified

pierreay opened the pull request:
## Description

By default, the `convert` plugin will skip every pending conversion if the converted file is already existing on the filesystem. However, when making changes (_e.g._, tags) in the original files, the changes are not reflected to the converted library. Adding the `--refresh` command-line option and the `refresh` option to the `convert` plugin allows to re-convert the original files if they are modified more recently than the time when converted files were converted.

## To Do

- [X] Documentation.
- [X] Changelog.
- [X] Tests.

Commit:
Implement the refresh behavior for convert command

Commit:
Implement the --refresh/-r command line option for convert's new feature

Commit:
Implement the convert.refresh configuration option

Commit:
Update documentation according for new convert --refresh option

Commit:
Update the changlog for new convert --refresh option

Commit:
Fix convert documentation paragraph order

Commit:
[beetsplug/convert] Fix linting error Ruff (E501)

Commit:
[beetsplug/convert] Fix linting error trailing whitespace

Commit:
[beetsplug/convert] Run ruff format

snejus commented on beetsplug/convert.py:
```diff
@@ -372,6 +384,29 @@ def convert_item(
+                    > os.path.getmtime(util.syspath(dest))
+                )
+            ):
+                if pretend:
+                    self._log.info(
+                        "rm {0} (original file modified)",
+                        util.displayable_path(dest),
+                    )
+                else:
+                    self._log.info(
+                        "Removing {0} (original file modified)",
+                        util.displayable_path(dest),
+                    )
+                    util.remove(util.syspath(dest))
```
You can simplify this

```suggestion
                self._log.info(
                    "Removing {0} (original file modified)",
                    util.displayable_path(dest),
                )
                if not pretend:
                    util.remove(dest)
```

snejus commented on beetsplug/convert.py:
```diff
@@ -372,6 +384,29 @@ def convert_item(
+            # --keep-new option because I'm not sure what to do in this case.
+            if (
+                (refresh and not keep_new)
+                and (os.path.exists(util.syspath(dest)))
+                and (
+                    os.path.getmtime(util.syspath(item.path))
+                    > os.path.getmtime(util.syspath(dest))
```
No need to call `util.syspath` here - both `item.path` and `item.destination` outputs are already in the expected form.

pierreay replied:
Indeed, my new commit `b96def17f` is removing all redundant `util.syspath` calls. ;)

snejus commented on beetsplug/convert.py:
```diff
@@ -372,6 +384,29 @@ def convert_item(
+            # Delete existing destination files when original files have been
+            # modified since the last conversion. NOTE: Only when not using the
+            # --keep-new option because I'm not sure what to do in this case.
+            if (
```
I think this may result in some unintended behaviour: see `should_transcode` check _below_: it seems that already existing encoded files will be removed regardless of this check (whether this music file should be transcoded)?

pierreay replied:
If I understand correctly, you mean that we should not remove a destination file because of the `--refresh` option if the original file should *not* be transcoded because it will be symlinked or hardlinked (thus, removing the need of performing the `convert` operation another time because of an original file being modified). Is that right? It yes, I think you're right and this can be improved, while not being harmful, isn't it?

pierreay replied:
Well, even if `should_transcode` is false, the file can be copied too (instead of linked). So that makes sense to remove `dest` such that the copy is done again when `refresh` is true from my understanding. What am I missing here?

snejus replied:
I think this looks fine now!

snejus commented on beetsplug/convert.py:
```diff
@@ -372,6 +384,29 @@ def convert_item(
+            # modified since the last conversion. NOTE: Only when not using the
+            # --keep-new option because I'm not sure what to do in this case.
+            if (
+                (refresh and not keep_new)
```
I think the number of duplicate checks is getting a bit too confusing: now `keep_new` is checked 3 times in 3 separate conditional branches, while existence of `dest` is checked twice. 

I think we want to combine/nest these.

pierreay replied:
Indeed, this was a bit messy. My last commit `a956a10bc` try to address this issue.

Commit:
style(convert.py): no newline

Co-authored-by: Šarūnas Nejus <snejus@protonmail.com>

Commit:
refactor(convert.py): simplify logic and redundant function

- Logic refactoring
- `util.syspath` is redundant because already called in `util.remove`

Co-authored-by: Šarūnas Nejus <snejus@protonmail.com>

Commit:
refactor(convert.py): remove uneeded util.syspath() calls

- `util.syspath` is redundant because already called in `util.normpath`
  itself called in Item class
Co-authored-by: Šarūnas Nejus <snejus@protonmail.com>

Commit:
refactor(convert): factorize multiple logic tests

Nest and combine some logic tests by:
1. Only check for destination file existence once.
2. Use `original` instead of `item.path` in the `refresh` conditional
   such that this option will be compatible with `keep_new`.

Commit:
fix(convert.py): fix missing syntax + format using ruff

Commit:
style(convert.py): fix ruff complain

pierreay commented:
Well, I just rebased on master anticipating a future merge, but I'm not sure the moment was right. ^^"

snejus commented:
Now the only missing bit is a note in the changelog as this is a new feature :)

pierreay committed:
docs(changelog): move --refresh entry in next release section

pierreay commented:
> Now the only missing bit is a note in the changelog as this is a new feature :)

@snejus Done! In fact the changelog entry was already there, but was lost in others entries of nearly one year ago since this is a pretty old PR ahah. :)

pierreay commented:
Rebased on master for `v2.3.0`. :) Anything else that I can do?

snejus commented on beetsplug/convert.py:
```diff
@@ -382,12 +393,24 @@ def convert_item(
+                        "Removing {0} (original file modified)",
+                        util.displayable_path(dest),
+                    )
+                    if not pretend:
```
Thanks for tidying this up. It's much more clear now, where the logic is clear.

The only question I have left: what happens when `pretend=True`?

As far as I can see, the file will not be removed and the logic will continue on line 415. Before, the logic would never reach line 415 if the file existed. 

Is this intentional?

snejus reviewed (commented):
Apologies for the delay!

JOJ0 commented:
Hi @pierreay, I'd like to ressurect this PR. There is a few suggestions from @snejus you might want to address, rebase against master and some polish and we should be good to go. I'll try to help move this forward. Thanks!

snejus commented:
Hey @pierreay, thanks for the detailed write-up and the diagram!

I think option 1 is the right call here. `keep_new` is already a niche feature, and adding `refresh` behavior on top of it would be a niche-of-niche code path that nobody has asked for yet. Since you yourself don't use `keep_new` and aren't sure whether it makes sense - that's a strong signal to not build it.

The current behavior (silently no-op) is harmless, so I'd say just make it explicit with an early guard and maybe a debug log message, e.g.:

```py
if keep_new:    
    self._log.debug(                                                                                                                                                              
        "Skipping refresh: not supported with keep_new"
    )
```

If a real user comes along later and needs scenario 2, we can build it then with a concrete use case to guide the implementation.

Let's keep the scope tight and get this merged - the PR has been open long enough!

snejus commented:
@pierreay just pinging to check whether you're still happy to finish this up?

pierreay commented:
> @pierreay just pinging to check whether you're still happy to finish this up?

Hi @snejus, yes I am! April and May were really busy but I will complete this PR very soon in the next days. ;)

(Should be done by the 22nd of june)

pierreay committed:
style(convert): Harmonize logging message when using pretend for refresh feature

- Also hamonize the logging message about moving files from `keep_new`
when using `pretend`
- According to snejus comment from 2025-07-06

pierreay committed:
refactor(convert): Better logic to handle both refresh and keep_new features

- Implement @snejus comment from the 2026-04-19
- This change the silent no-op when both refresh and keep_new are used
to proper logging messages depending on what combination of option is
used
- Also harmonize the usage of terms "target" and "destination" between
  logging messages

pierreay committed:
style(convert): Consistent usage of original vs. item.path variables

Except for the `keep_new` mechanism, it is not useful to use the `original`
variable which is equal to the `item.path` variable. Since refresh only
makes sense when `keep_new` is false, only use for the pair of `item.path`
and `dest` variables for clarity (and not `original` and `converted`
variables).

pierreay committed:
chore(convert): Poe format

pierreay commented:
- [x] Implement better logging
- [x] Implement better logic to handle `keep_new` and `refresh`
- [x] Rebase (lot of changes, take some time...)
- [x] Tests are passing
- [x] Final live test with a MWE of a database

<details><summary>Details</summary>
<p>
Should be finished by the 23
</p>
</details>

pierreay committed:
fix(convert): continue -> return

This continue comes from a rebase, when the convert_item were using a
while loop. Now we just have to go out of the function!

pierreay committed:
fix(convert): Get default value for refresh from config

pierreay committed:
chore(convert): Format

pierreay committed:
chore(docs): Format convert

pierreay committed:
style(convert): Simplify logic

This is a suggestion from ruff linting: get rid of the redundant `else`
after the `return`.

pierreay committed:
refactor(convert): Merge redundant checks

pierreay committed:
chore(changelog): Lint

pierreay commented:
Hello @snejus @JOJ0 ! I have been able to fix and implement all the things we discussed above, and rebasing on `master`. :) I think this time this PR should be ready to merge! (at least, hope so!)

snejus commented on beetsplug/convert.py:
```diff
@@ -400,7 +409,7 @@ def get_item_destination(self, item: Item) -> bytes:
 
     @pipeline.mutator_stage
-    def convert_item(self, keep_new: bool, item: Item) -> None:
+    def convert_item(self, refresh: bool, keep_new: bool, item: Item) -> None:
```
Define a new `@cached_property` called `refresh` and read the value from the config there. (See the rest of variables)

snejus reviewed (changes requested):
Looks good, just make `refresh` a cached property and access it directly from the methods. I passed `keep_new` around due to another reason, so it has to be passed in.

pierreay committed:
refactor(convert): Use cached_property for refresh option

- Fix snejus comment from 2026-06-23

pierreay commented:
> Looks good, just make `refresh` a cached property and access it directly from the methods. I passed `keep_new` around due to another reason, so it has to be passed in.

Nice catch, done! :)

snejus reviewed (approved):
Finally, this is ready and about to get merged! <3
